### DISCUSSION TITLE
How do I update the Java tests for a DITA-OT pull request?

### DISCUSSION THREAD

**Discussion by chrispy-snps** [dita-ot/dita-ot#4009] [Q&A]
I have enough XSLT experience to propose an occasional improvement to the DITA-OT. However, I have no Java programming experience, so I am not sure how to update unit tests accordingly.

Some questions I have are:

* Where are the tests stored, and how are they organized?
* How do I determine whether to modify an existing test or add a new test?
* How do I update the expected golden output for a test?
* How can I run just a specific test from `gradlew`, so I can iterate until I get things working correctly?

I can maybe figure out some of this by copying what is already there ("monkey-see, monkey-do!"), but some guidance in how to approach this process would be very helpful!

**Comment by raducoravu** [comment]
Here's my experience with creating tests for the DITA OT:

Let's say I have installed Eclipse, installed the Gradle plugin in it, imported the dita-ot project in Eclipse.
For each module class like "org.dita.dost.module.KeyrefModule" there is a corresponding test class like "org.dita.dost.module.KeyrefModuleTest".
For each test class there is a folder containing initial DITA sources "dita-ot/src/test/resources/KeyrefModuleTest/src" and a folder containing expected content, content resulting after the module has been applied by the test case in "dita-ot/src/test/resources/KeyrefModuleTest/exp"
A test case like the one I created here "org.dita.dost.module.KeyrefModuleTest.testWalkMapAndRewriteKeydefHrefKeepAnchor()"  usually attempts to load a file from the "KeyrefModuleTest/src" folder, applies the KeyrefModule on it and asserts some expected content.
Sometimes it takes me more to build the test case than to make the fix :)
And you need to know Java, without knowing Java making such a test is probably not feasible.

There are also a few testcases for XSLT stylesheets and templates, they are based on the XSpec framework and the only one I created so far was added in this XSpec file: /dost/src/test/xsl/preprocess/conrefImpl.xspec

**Reply by chrispy-snps** [reply]
@raducoravu - thanks for sharing this!

I was hoping that maybe writing a test was as simple as capturing some files before and after a preprocessing stage, but after looking at the files you referenced and the Java code that drives the tests, I see it is not quite so simple.

**Reply by raducoravu** [reply]
Yes, I also feel that the DITA OT is missing some kind of framework for fully black box tests.

**Comment by jelovirt** [comment]
>  Where are the tests stored, and how are they organized?

All the test code is under `src/test`. There are basically four types of tests

* unit tests that test smallish Java features
* module tests that test preprocess modules, usually they extend [AbstractModuleTest](https://github.com/dita-ot/dita-ot/blob/develop/src/test/java/org/dita/dost/module/AbstractModuleTest.java)
* integration tests that run the whole OT process and the test compares either preprocess output from temp directory or the output, depending on the tes
* end to end test that run some small conversion, even with transtypes we cannot compare to make sure the output is correct. These are regression tests to verify a given transtype can be run, regardless of output
* XSpec tests

>    How do I determine whether to modify an existing test or add a new test?

You really need to know what sort of tests there already are and decide based on that,

>    How do I update the expected golden output for a test?

It depends on the test. Usually the `src/test/resoures` has a `exp` directory for expected output. You update those files for your test and rerun the test.

>    How can I run just a specific test from gradlew, so I can iterate until I get things working correctly?

See e.g. https://stackoverflow.com/a/31468902
